samples/BenchmarkDotNet.Samples/IntroNuget.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using BenchmarkDotNet.Analysers;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Columns;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Engines;
+using BenchmarkDotNet.Exporters.Csv;
+using BenchmarkDotNet.Jobs;
+using BenchmarkDotNet.Loggers;
+using BenchmarkDotNet.Toolchains.CsProj;
+
+namespace BenchmarkDotNet.Samples
+
+    [Config(typeof(Config))]
+    public class IntroNuget
+    {
+        private class Config : ManualConfig
+        {
+            public Config()
+            {
+                //TODO: So far only implemented with any toolchain using DotNetCliBuilder
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "11.0.1").WithId("11.0.1"));
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "10.0.3").WithId("10.0.3"));
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "9.0.1").WithId("9.0.1"));
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "8.0.1").WithId("8.0.1"));
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "7.0.1").WithId("7.0.1"));
+                Add(Job.MediumRun.With(CsProjCoreToolchain.Current.Value).WithNuget("Newtonsoft.Json", "6.0.1").WithId("6.0.1"));
+                Add(DefaultConfig.Instance.GetColumnProviders().ToArray());
+                Add(DefaultConfig.Instance.GetLoggers().ToArray());
+                Add(CsvExporter.Default);
+            }
+        }
+
+        [GlobalSetup]
+        public void Setup()
```

Comment:
If the project that defines benchmarks (let's call it A) references `Newtonsoft.Json 9.0.1` and we are going to generate two projects:

1. 'Newtonsoft.Json 11.0.1' (B)
2. 'Newtonsoft.Json 6.0.1' (C)

Then which version of 'Newtonsoft.Json' is going to be restored and loaded during the execution? Will it be 11.0.1  for B and 9.0.1 for C? Is this why you are using reflection here?

Reply:
That's the million dollar question :) I'm using reflection in this test for 2 reasons: 1) so we don't have to reference this assembly in this project and 2) precisely to avoid this issue

I've been working to get Benchmark.Net to run benchmarks between many of our products' versions and have struggled with this very question. In an ideal world, you could do this:

* Project "A" (the benchmark project) references the minimum version of the nuget dependency that is being tested in the list of Jobs so that reflection doesn't have to be used to call the APIs and it builds
* The list of Jobs specifies the exact version that will be tested and that is what will be resolved from Nuget
* So long as the API signatures haven't changed between all of these versions and the assembly isn't strong-named (or there's no major version changes) then it should 'just work'

For simple projects/tests the above should be fine but of course ideal worlds don't always exist. In our case there's quite a bit of method signature changes between versions so this solution isn't going to work.  I've had a couple of ideas (some perhaps zany!): 

1. Have 'baseline' projects, one for each minimum nuget version to be tested for a particular API surface area. Then define the Jobs in these projects that reference the higher nuget version dependencies that you know will work. Then you could call these command line apps in sequence  and combine the csv results either manually or potentially with some sort of custom `IExporter`. Otherwise, you could have a parent command line app that calls these in sequence and if we could serialize/deserialize the `Summary` then we could combine them and send them to an exporter (i.e. #305 ) ... This is a pretty round about way to get this done though!
1. Have 'baseline' nuget packages that include the benchmarks to run then in your Jobs, you would specify the baseline nuget package which by default references the minimum nuget version to be tested for a given API surface area, and also include the specific Nuget version of the dependency to be tested. There's a problem with this approach though because there's no way to convert this to `BenchmarkRunInfo` that the `BenchmarkRunner` needs. Something would have to be created/modified to be able to resolve the `BenchmarkRunInfo.Type` and `BenchmarkRunInfo.BenchmarkCase` for each Job run and the runner would somehow still need to know about which cases it will be running.
1. Create a new `BenchmarkRunner.RunProjects` methods which would be similar to above but instead of building nuget packages containing the benchmarks, you could point the `BenchmarkRunner` at a list of csproj files containing benchmarks. This also suffers from the same problem listed above though.
1. Utilize something like the `RunUrl` (maybe `RunFiles`) where you point the `BenchmarkRunner` to a list of files, these files also contain metadata in a header (like yaml or something) which is parsed out and could contain assembly/nuget references. The `RunFiles` method has a parameter for the user to specify which benchmarks to run by name. Then Benchmark.Net builds the first file, finds the specified benchmarks by name which it uses as it's `BenchmarkRunInfo`, it executes that benchmark, then moves on to the next file, builds it, verifies it, runs it, and so on, all in the while collection a list of `Summary` objects which it can then combine at the end with `Summary.Join`

Sorry for very long message, but been meaning to write this all down somewhere! Let me know what you think or any ideas you've got.